src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,329 @@
+        :param command_args: Command name + arguments as a tuple
+        :type command_args: tuple
+        :param finished_hosts_key: Model attribute name for finished hosts
+        :type finished_hosts_key: str
+        """
+        # Split out the command name.
+        command_name = command_args[0]
+        command_args = command_args[1:]
+
+        end_status = 'finished'
+
+        # XXX We assume the model instance names a cluster.
+        #     Note, cluster_name is used in the except clause,
+        #     so it must be reliably defined.
+        cluster_name = getattr(model_instance, 'name', None)
+
+        try:
+            assert cluster_name is not None
+            model_json_data = model_instance.to_json()
+
+            # Set the initial status in the store.
+            self.logger.info('Setting initial status.')
+            self.logger.debug('Status={}'.format(model_json_data))
+            params = {
+                'model_type_name': model_instance.__class__.__name__,
+                'model_json_data': model_json_data
+            }
+            self.request('storage.save', params=params)
+
+            # Respond to the caller with the initial status.
+            if message.properties.get('reply_to'):
+                # XXX Have to dig up the message ID again.
+                #     CommissaireService.on_message() already
+                #     does this, but doesn't pass it to us.
+                body = message.body
+                if isinstance(body, bytes):
+                    body = json.loads(body.decode())
+                self.respond(
+                    message.properties['reply_to'],
+                    body.get('id', -1),
+                    model_json_data)
+        except Exception as error:
+            self.logger.error(
+                'Unable to save initial state for "{}" clusterexec due to '
+                '{}: {}'.format(cluster_name, type(error), error))
+            raise error
+
+        # Collect all host addresses in the cluster.
+
+        try:
+            cluster = Cluster.new(name=cluster_name)
+            params = {
+                'model_type_name': cluster.__class__.__name__,
+                'model_json_data': cluster.to_json(),
+                'secure': True
+            }
+            response = self.request('storage.get', params=params)
+            cluster = Cluster.new(**response['result'])
+        except Exception as error:
+            self.logger.warn(
+                'Unable to continue for cluster "{}" due to '
+                '{}: {}'.format(cluster_name, type(error), error))
+            raise error
+
+        n_hosts = len(cluster.hostset)
+        if n_hosts:
+            self.logger.debug(
+                '{} hosts in cluster "{}"'.format(n_hosts, cluster_name))
+        else:
+            self.logger.warn('No hosts in cluster "{}"'.format(cluster_name))
+
+        # TODO Find a better way to do this...
+
+        try:
```

Comment:
nit: I'm aware this is a direct port of the MVP portion but I feel like this could be inverted. Instead of getting a list of all hosts, iterating through each and checking if it is in the cluster's ``hostset``, couldn't we take the cluster's ``hostset`` and pull each host from storage? Seems like it would be less intensive.